File: src-lang/xt/db/node/example_auth_profile_ui.clj
(ns xt.db.node.example-auth-profile-ui
  (:require [hara.lang :as l]))

(l/script :js
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]
             [xt.lang.spec-promise :as promise]
             [xt.substrate.page-core :as page-core]
             [xt.db.node.example-auth-profile :as example]
             [js.react :as r]]})

(defn.js error-message
  "returns a readable message for an async auth failure"
  {:added "4.1"}
  [err]
  (return (or (xt/x:ex-message err)
              (xt/x:to-string err))))

(defn.js refresh-auth-state
  "refreshes the session/profile models and updates the supplied React setters"
  {:added "4.1"}
  [node space-id group-id setSession setProfile]
  (return
   (-> (page-core/model-refresh node space-id group-id "session" {} nil)
       (promise/x:promise-then
        (fn [_]
          (var curr-session
               (page-core/model-get-output node space-id group-id "session"))
          (setSession curr-session)
          (if (xt/x:not-nil? curr-session)
            (return
             (-> (page-core/model-refresh node space-id group-id "profile" {} nil)
                 (promise/x:promise-then
                  (fn [_]
                    (var profile-output
                         (page-core/model-get-output node space-id group-id "profile"))
                    (setProfile (xtd/get-in profile-output ["user"]))
                    (return profile-output)))))
            (do (setProfile nil)
                (return nil))))))))

(defn.js run-action
  "runs a page-model action and refreshes the visible auth state"
  {:added "4.1"}
  [node space-id group-id model-id event
   setBusy setError setSession setProfile]
  (setBusy model-id)
  (setError nil)
  (return
   (-> (page-core/model-refresh node
                                space-id
                                group-id
                                model-id
                                (or event {})
                                nil)
       (promise/x:promise-then
        (fn [_]
          (return
           (-/refresh-auth-state node
                                 space-id
                                 group-id
                                 setSession
                                 setProfile))))
       (promise/x:promise-then
        (fn [out]
          (setBusy nil)
          (return out)))
       (promise/x:promise-catch
        (fn [err]
          (setBusy nil)
          (setError (-/error-message err))
          (return nil))))))

(defn.js field-style
  []
  (return {:display "flex"
           :flexDirection "column"
           :gap "6px"
           :fontSize "13px"
           :fontWeight 600
           :color "#374151"}))

(defn.js input-style
  []
  (return {:border "1px solid #d1d5db"
           :borderRadius "8px"
           :padding "10px 12px"
           :fontSize "14px"
           :fontWeight 400
           :outline "none"}))

(defn.js button-style
  [primary]
  (return {:border "1px solid"
           :borderColor (:? primary "#111827" "#d1d5db")
           :background (:? primary "#111827" "#ffffff")
           :color (:? primary "#ffffff" "#111827")
           :borderRadius "8px"
           :padding "9px 13px"
           :fontSize "13px"
           :fontWeight 600
           :cursor "pointer"}))

(defn.js AuthProfileApp
  "interactive auth/profile example backed by the attached page models"
  {:added "4.1"}
  [#{node space-id group-id initial-email}]
  (:= space-id (or space-id example/DEFAULT_SPACE_ID))
  (:= group-id (or group-id example/DEFAULT_GROUP_ID))
  (var [email setEmail]
       (r/local
        (or initial-email
            (xt/x:cat "auth-profile-"
                      (xt/x:to-string (xt/x:now-ms))
                      "@example.com"))))
  (var [password setPassword] (r/local "secret123"))
  (var [displayName setDisplayName] (r/local "Playground User"))
  (var [session setSession] (r/local nil))
  (var [profile setProfile] (r/local nil))
  (var [busy setBusy] (r/local nil))
  (var [error setError] (r/local nil))
  (var credentials {"email" email
                    "password" password})
  (var current-user (or profile
                        (xt/x:get-key session "user")))
  (var current-email (xt/x:get-key current-user "email"))
  (var current-name (xtd/get-in current-user
                                ["user_metadata" "display_name"]))
  (var signed-in (xt/x:not-nil? session))
  (r/init []
    (-/refresh-auth-state node
                          space-id
                          group-id
                          setSession
                          setProfile)
    (return nil))
  (return
   [:main
    {:style {:minHeight "100%"
             :boxSizing "border-box"
             :padding "32px"
             :background "#f3f4f6"
             :fontFamily "system-ui, sans-serif"
             :color "#111827"}}
    [:div
     {:style {:maxWidth "720px"
              :margin "0 auto"
              :display "flex"
              :flexDirection "column"
              :gap "18px"}}
     [:header
      [:div
       {:style {:fontSize "12px"
                :fontWeight 700
                :letterSpacing "0.08em"
                :textTransform "uppercase"
                :color "#6b7280"}}
       "foundation-base / local-min"]
      [:h1
       {:style {:margin "6px 0 4px"
                :fontSize "30px"}}
       "Supabase auth profile"]
      [:p
       {:style {:margin 0
                :color "#4b5563"
                :lineHeight 1.5}}
       "Create or sign in to a local account, then update the user's auth metadata through substrate page models."]]
     [:section
      {:style {:background "#ffffff"
               :border "1px solid #e5e7eb"
               :borderRadius "12px"
               :padding "20px"
               :display "flex"
               :flexDirection "column"
               :gap "14px"
               :boxShadow "0 8px 24px rgba(17,24,39,0.05)"}}
      [:label
       {:style (-/field-style)}
       "Email"
       [:input
        {:style (-/input-style)
         :value email
         :disabled signed-in
         :onChange (fn [event]
                     (setEmail (. event target value)))}]]
      [:label
       {:style (-/field-style)}
       "Password"
       [:input
        {:style (-/input-style)
         :type "password"
         :value password
         :disabled signed-in
         :onChange (fn [event]
                     (setPassword (. event target value)))}]]
      [:div
       {:style {:display "flex"
                :gap "10px"
                :flexWrap "wrap"}}
       [:button
        {:style (-/button-style true)
         :disabled (xt/x:not-nil? busy)
         :onClick (fn []
                    (-/run-action node
                                  space-id
                                  group-id
                                  "sign-up"
                                  {"credentials" credentials}
                                  setBusy
                                  setError
                                  setSession
                                  setProfile))}
        (:? (== busy "sign-up")
            "Creating..."
            "Create account")]
       [:button
        {:style (-/button-style false)
         :disabled (xt/x:not-nil? busy)
         :onClick (fn []
                    (-/run-action node
                                  space-id
                                  group-id
                                  "login"
                                  {"credentials" credentials}
                                  setBusy
                                  setError
                                  setSession
                                  setProfile))}
        (:? (== busy "login")
            "Signing in..."
            "Sign in")]
       [:button
        {:style (-/button-style false)
         :disabled (or (xt/x:not-nil? busy)
                       (not signed-in))
         :onClick (fn []
                    (-/run-action node
                                  space-id
                                  group-id
                                  "logout"
                                  {}
                                  setBusy
                                  setError
                                  setSession
                                  setProfile))}
        (:? (== busy "logout")
            "Signing out..."
            "Sign out")]]
      (:? error
          [:div
           {:style {:padding "10px 12px"
                    :borderRadius "8px"
                    :background "#fef2f2"
                    :color "#b91c1c"
                    :fontSize "13px"}}
           error]
          nil)]
     [:section
      {:style {:background "#ffffff"
               :border "1px solid #e5e7eb"
               :borderRadius "12px"
               :padding "20px"
               :display "flex"
               :flexDirection "column"
               :gap "14px"}}
      [:div
       {:style {:display "flex"
                :justifyContent "space-between"
                :gap "16px"
                :alignItems "flex-start"}}
       [:div
        [:h2
         {:style {:margin "0 0 4px"
                  :fontSize "18px"}}
         "Current session"]
        [:div
         {:style {:color "#6b7280"
                  :fontSize "13px"}}
         (:? signed-in
             (or current-email "Authenticated user")
             "Not signed in")]]
       [:span
        {:style {:padding "5px 9px"
                 :borderRadius "999px"
                 :fontSize "12px"
                 :fontWeight 700
                 :background (:? signed-in "#dcfce7" "#f3f4f6")
                 :color (:? signed-in "#166534" "#6b7280")}}
        (:? signed-in "signed in" "signed out")]]
      [:label
       {:style (-/field-style)}
       "Display name"
       [:input
        {:style (-/input-style)
         :value displayName
         :disabled (not signed-in)
         :onChange (fn [event]
                     (setDisplayName (. event target value)))}]]
      [:button
       {:style (-/button-style true)
        :disabled (or (xt/x:not-nil? busy)
                      (not signed-in))
        :onClick (fn []
                   (-/run-action node
                                 space-id
                                 group-id
                                 "change-profile"
                                 {"profile" {"display_name" displayName}}
                                 setBusy
                                 setError
                                 setSession
                                 setProfile))}
       (:? (== busy "change-profile")
           "Saving..."
           "Save profile")]
      (:? current-name
          [:div
           {:style {:fontSize "13px"
                    :color "#374151"}}
           "Stored display name: "
           [:strong current-name]]
          nil)]]]))

(defn.js render-playground
  "renders the auth/profile UI into the active playground stage"
  {:added "4.1"}
  [node opts]
  (:= opts (or opts {}))
  (var space-id (or (xt/x:get-key opts "space_id")
                    example/DEFAULT_SPACE_ID))
  (var group-id (or (xt/x:get-key opts "group_id")
                    example/DEFAULT_GROUP_ID))
  (window.PLAYGROUND.setTitle "Supabase auth profile")
  (window.PLAYGROUND.setStage
   [:% -/AuthProfileApp
    {:node node
     :space-id space-id
     :group-id group-id
     :initial-email (xt/x:get-key opts "email")}])
  (return node))

(defn.js mount-playground
  "creates the example node and mounts its UI in a :runtime :playground page"
  {:added "4.1"}
  [client-defaults opts]
  (:= opts (or opts {}))
  (var service-id (or (xt/x:get-key opts "service_id")
                      example/DEFAULT_SERVICE_ID))
  (var page-args {"space_id" (or (xt/x:get-key opts "space_id")
                                  example/DEFAULT_SPACE_ID)
                  "group_id" (or (xt/x:get-key opts "group_id")
                                  example/DEFAULT_GROUP_ID)})
  (var node (example/create-auth-profile-node client-defaults
                                               service-id
                                               page-args))
  (-/render-playground node opts)
  (return node))

File: packages-gen/js/libs/xt-db/node/example-auth-profile-ui.js
const React = require("react")

const example = require("@xtalk/db/node/example-auth-profile.js")

const xtd = require("@xtalk/lang/common-data.js")

const view_runtime = require("@xtalk/db/react/view/runtime.js")

const page_core = require("@xtalk/substrate/page-core.js")

const shared_view = require("@xtalk/db/node/example-auth-profile-view.js")

function error_message(err){
  return ((err instanceof Error) ? err["message"] : null) || String(err);
}

function refresh_auth_state(node,space_id,group_id,setSession,setProfile){
  return page_core.model_refresh(node,space_id,group_id,"session",{},null).then(function (_){
    let curr_session = page_core.model_get_output(node,space_id,group_id,"session");
    setSession(curr_session);
    if(null != curr_session){
      return page_core.model_refresh(node,space_id,group_id,"profile",{},null).then(function (_){
        let profile_output = page_core.model_get_output(node,space_id,group_id,"profile");
        setProfile(xtd.get_in(profile_output,["user"]));
        return profile_output;
      });
    }
    else{
      setProfile(null);
      return null;
    }
  });
}

function run_action(node,space_id,group_id,model_id,event,setBusy,setError,setSession,setProfile){
  setBusy(model_id);
  setError(null);
  return page_core.model_refresh(node,space_id,group_id,model_id,event || {},null).then(function (_){
    return refresh_auth_state(node,space_id,group_id,setSession,setProfile);
  }).then(function (out){
    setBusy(null);
    return out;
  }).catch(function (err){
    setBusy(null);
    setError(error_message(err));
    return null;
  });
}

function field_style(){
  return {
    "display":"flex",
    "flexDirection":"column",
    "gap":"6px",
    "fontSize":"13px",
    "fontWeight":600,
    "color":"#374151"
  };
}

function input_style(){
  return {
    "border":"1px solid #d1d5db",
    "borderRadius":"8px",
    "padding":"10px 12px",
    "fontSize":"14px",
    "fontWeight":400,
    "outline":"none"
  };
}

function button_style(primary){
  return {
    "color":primary ? "#ffffff" : "#111827",
    "borderRadius":"8px",
    "borderColor":primary ? "#111827" : "#d1d5db",
    "background":primary ? "#111827" : "#ffffff",
    "cursor":"pointer",
    "padding":"9px 13px",
    "fontWeight":600,
    "fontSize":"13px",
    "border":"1px solid"
  };
}

function AuthProfileApp({group_id,initial_email,node,space_id}){
  space_id = (space_id || example.DEFAULT_SPACE_ID);
  group_id = (group_id || example.DEFAULT_GROUP_ID);
  let [email,setEmail] = React.useState(
    initial_email || ("auth-profile-" + String(Date.now()) + "@example.com")
  );
  let [password,setPassword] = React.useState("secret123");
  let [displayName,setDisplayName] = React.useState("Playground User");
  let [session,setSession] = React.useState();
  let [profile,setProfile] = React.useState();
  let [busy,setBusy] = React.useState();
  let [error,setError] = React.useState();
  let credentials = {"email":email,"password":password};
  let current_user = profile || session["user"];
  let current_email = current_user["email"];
  let current_name = xtd.get_in(current_user,["user_metadata","display_name"]);
  let signed_in = null != session;
  React.useEffect(function (){
    refresh_auth_state(node,space_id,group_id,setSession,setProfile);
    return null;
  },[]);
  return (
    <main
      style={{
          "minHeight":"100%",
          "boxSizing":"border-box",
          "padding":"32px",
          "background":"#f3f4f6",
          "fontFamily":"system-ui, sans-serif",
          "color":"#111827"
        }}>
      <div
        style={{
            "maxWidth":"720px",
            "margin":"0 auto",
            "display":"flex",
            "flexDirection":"column",
            "gap":"18px"
          }}>
        <header>
          <div
            style={{
                "fontSize":"12px",
                "fontWeight":700,
                "letterSpacing":"0.08em",
                "textTransform":"uppercase",
                "color":"#6b7280"
              }}>foundation-base / local-min
          </div>
          <h1 style={{"margin":"6px 0 4px","fontSize":"30px"}}>Supabase auth profile</h1>
          <p style={{"margin":0,"color":"#4b5563","lineHeight":1.5}}>
            Create or sign in to a local account, then update the user's auth metadata through substrate page models.
          </p>
        </header>
        <section
          style={{
              "background":"#ffffff",
              "border":"1px solid #e5e7eb",
              "borderRadius":"12px",
              "padding":"20px",
              "display":"flex",
              "flexDirection":"column",
              "gap":"14px",
              "boxShadow":"0 8px 24px rgba(17,24,39,0.05)"
            }}>
          <label style={field_style()}>
            Email
            <input
              style={input_style()}
              value={email}
              disabled={signed_in}
              onChange={function (event){
                  setEmail(event.target.value);
                }}/>
          </label>
          <label style={field_style()}>
            Password
            <input
              style={input_style()}
              type="password"
              value={password}
              disabled={signed_in}
              onChange={function (event){
                  setPassword(event.target.value);
                }}/>
          </label>
          <div style={{"display":"flex","gap":"10px","flexWrap":"wrap"}}>
            <button
              style={button_style(true)}
              disabled={(null != busy) || signed_in}
              onClick={function (){
                  run_action(
                    node,
                    space_id,
                    group_id,
                    "sign-up",
                    {"credentials":credentials},
                    setBusy,
                    setError,
                    setSession,
                    setProfile
                  );
                }}>{(busy == "sign-up") ? "Creating..." : "Create account"}
            </button>
            <button
              style={button_style(false)}
              disabled={(null != busy) || signed_in}
              onClick={function (){
                  run_action(
                    node,
                    space_id,
                    group_id,
                    "login",
                    {"credentials":credentials},
                    setBusy,
                    setError,
                    setSession,
                    setProfile
                  );
                }}>{(busy == "login") ? "Signing in..." : "Sign in"}
            </button>
            <button
              style={button_style(false)}
              disabled={(null != busy) || !signed_in}
              onClick={function (){
                  run_action(
                    node,
                    space_id,
                    group_id,
                    "logout",
                    {},
                    setBusy,
                    setError,
                    setSession,
                    setProfile
                  );
                }}>{(busy == "logout") ? "Signing out..." : "Sign out"}
            </button>
          </div>
          {error ? (
            <div
              style={{
                  "padding":"10px 12px",
                  "borderRadius":"8px",
                  "background":"#fef2f2",
                  "color":"#b91c1c",
                  "fontSize":"13px"
                }}>{error}
            </div>) : null}
        </section>
        <section
          style={{
              "background":"#ffffff",
              "border":"1px solid #e5e7eb",
              "borderRadius":"12px",
              "padding":"20px",
              "display":"flex",
              "flexDirection":"column",
              "gap":"14px"
            }}>
          <div
            style={{
                "display":"flex",
                "justifyContent":"space-between",
                "gap":"16px",
                "alignItems":"flex-start"
              }}>
            <div>
              <h2 style={{"margin":"0 0 4px","fontSize":"18px"}}>Current session</h2>
              <div style={{"color":"#6b7280","fontSize":"13px"}}>
                {signed_in ? (current_email || "Authenticated user") : "Not signed in"}
              </div>
            </div>
            <span
              style={{
                  "padding":"5px 9px",
                  "borderRadius":"999px",
                  "fontSize":"12px",
                  "fontWeight":700,
                  "background":signed_in ? "#dcfce7" : "#f3f4f6",
                  "color":signed_in ? "#166534" : "#6b7280"
                }}>{signed_in ? "signed in" : "signed out"}
            </span>
          </div>
          <label style={field_style()}>
            Display name
            <input
              style={input_style()}
              value={displayName}
              disabled={!signed_in}
              onChange={function (event){
                  setDisplayName(event.target.value);
                }}/>
          </label>
          <button
            style={button_style(true)}
            disabled={(null != busy) || !signed_in}
            onClick={function (){
                run_action(
                  node,
                  space_id,
                  group_id,
                  "change-profile",
                  {"profile":{"display_name":displayName}},
                  setBusy,
                  setError,
                  setSession,
                  setProfile
                );
              }}>{(busy == "change-profile") ? "Saving..." : "Save profile"}
          </button>
          {current_name ? (
            <div style={{"fontSize":"13px","color":"#374151"}}>Stored display name: <strong>{current_name}</strong></div>) : null}
        </section>
      </div>
    </main>);
}

function render_playground(node,opts){
  opts = (opts || {});
  let space_id = opts["space_id"] || example.DEFAULT_SPACE_ID;
  let group_id = opts["group_id"] || example.DEFAULT_GROUP_ID;
  window.PLAYGROUND.setTitle("Supabase auth profile");
  window.PLAYGROUND.setStage((
    <AuthProfileApp
      node={node}
      spaceId={space_id}
      groupId={group_id}
      initialEmail={opts["email"]}/>));
  return node;
}

function render_substrate_view_playground(node,opts){
  opts = (opts || {});
  let spec = shared_view.install(node,opts);
  let space_id = opts["space_id"] || example.DEFAULT_SPACE_ID;
  window.PLAYGROUND.setTitle("Substrate auth profile");
  window.PLAYGROUND.setStage((
    <view_runtime.View
      node={node}
      spec={spec}
      renderFn={shared_view.render}
      options={{"space_id":space_id}}/>));
  return node;
}

function mount_playground(client_defaults,opts){
  opts = (opts || {});
  let service_id = opts["service_id"] || example.DEFAULT_SERVICE_ID;
  let page_args = {
    "space_id":opts["space_id"] || example.DEFAULT_SPACE_ID,
    "group_id":opts["group_id"] || example.DEFAULT_GROUP_ID
  };
  let node = example.create_auth_profile_node(client_defaults,service_id,page_args);
  render_playground(node,opts);
  return node;
}

module.exports = {
  ["error_message"]:error_message,
  ["refresh_auth_state"]:refresh_auth_state,
  ["run_action"]:run_action,
  ["field_style"]:field_style,
  ["input_style"]:input_style,
  ["button_style"]:button_style,
  ["AuthProfileApp"]:AuthProfileApp,
  ["render_playground"]:render_playground,
  ["render_substrate_view_playground"]:render_substrate_view_playground,
  ["mount_playground"]:mount_playground
}